### DISCUSSION TITLE
Add procedural prototype textures

### DISCUSSION THREAD

**Discussion by RedMser** [godotengine/godot-proposals#5294] [3D]
Especially in 3D (but also valid for 2D), icon.png and noisetexture just isn't enough to create a quick prototype with distinct surfaces, even when combined with modulate colors.

Instead of bundling different prototype textures with the engine, why not create new texture types which **procedurally** generate them. I suggest the following:

![image](https://user-images.githubusercontent.com/5117197/187498468-9cd73dcf-55bf-42e1-aee0-9d49dfc03659.png)

- CheckerTexture
  - Generates a checkerboard texture
  - Allow setting two colors and a scaling factor

---
![image](https://user-images.githubusercontent.com/5117197/187498563-700c7ee4-12a4-496e-98ff-788db4d66f5b.png)

- GridTexture
  - Generates grid lines
  - Allow setting background color and grid color, as well as a scaling factor
  - Maybe support major/minor grid lines as well (possibly by using an opacity slider)

**Comment by dalexeev** [comment]
Another possible option is to display the size of the texture:

![](https://user-images.githubusercontent.com/47700418/187604679-00c841dc-ef74-4149-8da1-790372cff06f.png)

**Comment by TamerSoup625** [comment]
An addon could be created for this.
You can create procedural textures using `Image` resource and `ImageTexture`.

**Reply by dalexeev** [reply]
> An addon could be created for this.
> You can create procedural textures using `Image` resource and `ImageTexture`.

[dalexeev/godot-test-texture](https://github.com/dalexeev/godot-test-texture)

**Reply by timothyqiu** [reply]
Without [global addon](https://github.com/godotengine/godot-proposals/issues/831) implemented, I think this somehow defeats the purpose of procedural generation.

Installing an addon is no different from copying a bunch of premade textures.

**Comment by seacat17** [comment]
This can be used not only for testing, but also for saving some disk space. What if I want to make a floor with checkerboard pattern? I would just generate the texture on-the-fly

**Comment by passivestar** [comment]
This will be helpful not only for game prototypes but also for tech demos and issue reports. Because these can easily be added to MRPs and screenshots (which is an argument for having it in core and not as an addon)

**Comment by Calinou** [comment]
I made a proof of concept with a GridTexture class that can do both grid and checkers: https://github.com/Calinou/godot/tree/add-gridtexture

![gridtexture webp](https://github.com/godotengine/godot-proposals/assets/180032/37ba6c6d-e9d9-4b3b-ba65-5fcdc438b9c3)

**Testing project:** [test_grid_texture.zip](https://github.com/godotengine/godot-proposals/files/15333879/test_grid_texture.zip) (requires the branch linked above)

Feel free to continue this work if you'd like. Some outstanding things remain to be done:

- Support checker alternating colors.
- Support transparent grid colors when drawing the image to make the grid less visible or invisible.
- Support grid line thickness greater than `1` (the property currently doesn't do anything).
- Support primary grid lines to make lines thicker every N iterations (the property currently doesn't do anything).

The rest of the features seem to be working already.

Some happy accidents during development:

![gridtexture_1 webp](https://github.com/godotengine/godot-proposals/assets/180032/a144b4bd-3125-4e9f-9c82-22110eaf19e6)

![gridtexture_2 webp](https://github.com/godotengine/godot-proposals/assets/180032/036c92d3-a195-4ffe-8e24-e534f2870fd5)
